chat/views/auth.py

```diff
@@ -1,61 +1,135 @@
-from flask import Blueprint, g, render_template, request, session, redirect, current_app
-from flask.ext.openid import OpenID
+from flask import url_for, current_app, redirect, request, Blueprint, g, render_template, session
+from rauth import OAuth2Service
 from hashlib import md5
 from chat.datastore import db, add_user_to_channel, get_user
 from chat.zmq_context import zmq_context, push_socket
 from chat.views.eventhub import send_join_channel
-import urllib
+import urllib, urllib2
 import uuid
+import json
 
 auth = Blueprint("auth", __name__)
-oid = OpenID()
+
+class OAuthSignIn(object):
+  providers = None
+
+  def __init__(self, provider_name):
+    self.provider_name = provider_name
+    credentials = current_app.config['OAUTH_CREDENTIALS'][provider_name]
+    self.consumer_id = credentials['id']
+    self.consumer_secret = credentials['secret']
+
+  def authorize(self):
+    pass
+
+  def callback(self):
+    pass
+
+  def get_callback_url(self):
+    return url_for('auth.oauth_callback', provider=self.provider_name,
+                    _external=True)
+
+  @classmethod
+  def get_provider(self, provider_name):
+    if self.providers is None:
+      self.providers={}
+      for provider_class in self.__subclasses__():
+        provider = provider_class()
+        self.providers[provider.provider_name] = provider
+    return self.providers[provider_name]
+
+class GoogleSignIn(OAuthSignIn):
+  def __init__(self):
+    super(GoogleSignIn, self).__init__('google')
+    googleinfo = urllib2.urlopen('https://accounts.google.com/.well-known/openid-configuration')
+    google_params = json.load(googleinfo)
+    self.service = OAuth2Service(
+      name='google',
+      client_id=self.consumer_id,
+      client_secret=self.consumer_secret,
+      authorize_url=google_params.get('authorization_endpoint'),
+      base_url=google_params.get('userinfo_endpoint'),
+      access_token_url=google_params.get('token_endpoint')
+    )
+
+  def authorize(self):
+    return redirect(self.service.get_authorize_url(
+      scope='email profile',
+      response_type='code',
+      redirect_uri=self.get_callback_url())
+    )
+
+  def callback(self):
+    if 'code' not in request.args:
+      return None, None, None
+    oauth_session = self.service.get_auth_session(
+      data={'code': request.args['code'],
+            'grant_type': 'authorization_code',
+            'redirect_uri': self.get_callback_url()
+           },
+      decoder = json.loads
+    )
+    me = oauth_session.get('').json()
+    return (me['name'],
+            me['email'])
+
+@auth.route('/authorize/<provider>')
+def oauth_authorize(provider):
+  # Flask-Login function
+  if g.authed is True:
+    return redirect(url_for('frontend.index'))
+  print provider
```

Comment:
I don't have my machine with a working prat env with me.  I can do a fast follow PR tonight to clean up the spurious debug prints.